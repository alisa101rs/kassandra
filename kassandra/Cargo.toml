[package]
name = "kassandra"
version.workspace = true
authors.workspace = true
description = "Single node kassandra implementation"
repository.workspace = true
license.workspace = true
readme.workspace = true
edition.workspace = true
keywords.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tokio = { version = "1", features = ["macros"], default-features = false }
tokio-util = { version = "0.7.8", features = ["codec"] }
eyre = "0.6.8"
futures = "0.3.28"
futures-util = { version = "0.3.28", features = ["sink"] }
nom = "7.1.3"
bytes = { version = "1.4.0", features = ["serde"] }
bitflags = "2"
indexmap = { version = "2.1.0", features = ["serde"] }
seahash = "4.1.0"
thiserror = "1.0.40"
ulid = "1.0.0"
uuid = { version = "1.10.0", features = ["serde"] }
serde = { version = "1", features = ["derive"] }
serde_bytes = "0.11.11"
serde_json = "1.0.107"
derive_more = "1.0.0"
chrono = { version = "0.4.27", features = ["serde"] }
bigdecimal = { version = "0.4.0", features = ["serde"] }
num-bigint = { version = "0.4.3", features = ["serde"] }
num_enum = { version = "0.6" }
ron = "0.8.0"
strum = { version = "0.25", features = ["derive"] }
integer-encoding = "4.0.0"

tracing = "0.1"
tracing-subscriber = { version = "0.3.17", features = ["env-filter"] }

[dev-dependencies]
insta = { version = "1.34.0" }
