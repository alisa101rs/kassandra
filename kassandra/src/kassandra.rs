use std::collections::HashMap;

use bytes::Bytes;
use eyre::Result;
use serde::Serialize;
use tracing::{instrument, Level};

use crate::{
    cql::query::QueryString,
    error::DbError,
    frame::{
        request::{
            batch::{Batch, BatchStatement},
            execute::Execute,
            query::Query,
        },
        response::{
            error::Error,
            event::{SchemaChangeEvent, SchemaChangeType},
            result::{Prepared, QueryResult, SchemaChange, SetKeyspace},
        },
    },
    snapshot::DataSnapshots,
    storage::{
        keyspace::{Keyspace, Strategy},
        system::{system_keyspace, system_schema_keyspace},
    },
};

#[derive(Debug, Clone, Serialize)]
pub struct Kassandra {
    use_keyspace: Option<String>,
    keyspace: HashMap<String, Keyspace>,
    #[serde(skip)]
    prepared: HashMap<u128, QueryString>,
}

impl Default for Kassandra {
    fn default() -> Self {
        Self::new()
    }
}

impl Kassandra {
    pub fn new() -> Self {
        let keyspace = [system_keyspace(), system_schema_keyspace()]
            .into_iter()
            .collect();

        Self {
            use_keyspace: None,
            prepared: HashMap::new(),
            keyspace,
        }
    }

    pub fn load_state<'a>(data: impl IntoIterator<Item = &'a [u8]>) -> Result<Self> {
        let mut kassandra = Self::new();

        for piece in data {
            let k = ron::de::from_bytes::<HashMap<String, Keyspace>>(piece)?;
            kassandra.keyspace.extend(k.into_iter());
        }

        Ok(kassandra)
    }

    #[instrument(level = Level::TRACE, skip(self), err, ret)]
    pub fn process(&mut self, query: Query<'_>) -> Result<QueryResult, Error> {
        match &query.query {
            QueryString::Use { keyspace, .. } => {
                self.use_keyspace = Some(keyspace.clone());

                Ok(QueryResult::SetKeyspace(SetKeyspace {
                    keyspace_name: keyspace.to_owned(),
                }))
            }
            QueryString::Select { keyspace, .. } => {
                let Some(keyspace) = keyspace.as_deref().or(self.use_keyspace.as_deref()) else {
                    Err(Error::new(DbError::Invalid, "keyspace must be specified"))?
                };
                let Some(keyspace) = self.keyspace.get_mut(keyspace) else {
                    Err(Error::new(DbError::Invalid, "keyspace does not exist"))?
                };
                Ok(keyspace.select(query)?)
            }
            QueryString::Insert { keyspace, .. } => {
                let Some(keyspace) = keyspace.as_deref().or(self.use_keyspace.as_deref()) else {
                    Err(Error::new(DbError::Invalid, "keyspace must be specified"))?
                };
                let Some(keyspace) = self.keyspace.get_mut(keyspace) else {
                    Err(Error::new(DbError::Invalid, "keyspace does not exist"))?
                };
                Ok(keyspace.insert(query)?)
            }
            QueryString::Delete { keyspace, .. } => {
                let Some(keyspace) = keyspace.as_deref().or(self.use_keyspace.as_deref()) else {
                    Err(Error::new(DbError::Invalid, "keyspace must be specified"))?
                };
                let Some(keyspace) = self.keyspace.get_mut(keyspace) else {
                    Err(Error::new(DbError::Invalid, "keyspace does not exist"))?
                };
                Ok(keyspace.delete(query)?)
            }

            QueryString::CreateKeyspace { .. } => Ok(self.create_keyspace(query)?),
            QueryString::CreateTable { keyspace, .. } => {
                let Some(keyspace) = keyspace.as_deref().or(self.use_keyspace.as_deref()) else {
                    Err(DbError::Invalid)?
                };
                let Some(keyspace) = self.keyspace.get_mut(keyspace) else {
                    Err(DbError::Invalid)?
                };

                Ok(keyspace.create_table(query)?)
            }
            QueryString::CreateType { .. } => {
                unimplemented!()
            }
        }
    }

    #[instrument(level = Level::TRACE, skip(self), err, ret)]
    pub fn execute(&mut self, execute: Execute<'_>) -> Result<QueryResult, Error> {
        let Execute {
            id,
            consistency,
            flags,
            data,
            page_size,
            paging_state,
            serial_consistency,
            default_timestamp,
        } = execute;
        let parsed_id = u128::from_be_bytes(id.try_into().unwrap());
        let Some(query) = self.prepared.get(&parsed_id).cloned() else {
            return Err(Error::new(
                DbError::Unprepared {
                    statement_id: Bytes::copy_from_slice(id)
                },
                "unprepared query"
            ))
        };

        let query = Query {
            query,
            raw_query: "",
            consistency,
            flags,
            data,
            page_size,
            paging_state,
            serial_consistency,
            default_timestamp,
        };

        self.process(query)
    }

    #[instrument(level = Level::TRACE, skip(self), err, ret)]
    pub fn process_batch(&mut self, batch: Batch<'_>) -> Result<QueryResult, Error> {
        let Batch {
            batch_type: _,
            consistency: _,
            serial_consistency: _,
            timestamp: _,
            values,
        } = batch;

        for query in values {
            match query {
                BatchStatement::Query(query) => {
                    let QueryResult::Void = self.process(query)? else {
                        unimplemented!("should be only void")
                    };
                }
                BatchStatement::Prepared(execute) => {
                    let QueryResult::Void = self.execute(execute)? else {
                        unimplemented!("should be only void")
                    };
                }
            }
        }

        Ok(QueryResult::Void)
    }

    fn create_keyspace(&mut self, query: Query<'_>) -> Result<QueryResult, DbError> {
        let QueryString::CreateKeyspace { keyspace,ignore_existence,  ..} = query.query else {
            unreachable!()
        };
        if self.keyspace.contains_key(&keyspace) {
            return if ignore_existence {
                Ok(QueryResult::Void)
            } else {
                Err(DbError::AlreadyExists {
                    keyspace,
                    table: "".to_string(),
                })
            };
        }

        self.keyspace.insert(
            keyspace.clone(),
            Keyspace {
                name: keyspace.clone(),
                strategy: Strategy::LocalStrategy,
                tables: Default::default(),
                user_defined_types: Default::default(),
            },
        );

        let event = SchemaChangeEvent::KeyspaceChange {
            change_type: SchemaChangeType::Created,
            keyspace_name: keyspace,
        };

        Ok(QueryResult::SchemaChange(SchemaChange { event }))
    }

    pub fn save_state(&self) -> Vec<u8> {
        ron::ser::to_string_pretty(&self.keyspace, Default::default())
            .unwrap()
            .into_bytes()
    }

    pub fn data_snapshot(&self) -> DataSnapshots {
        DataSnapshots::from_keyspaces(&self.keyspace)
    }

    #[instrument(level = Level::TRACE, skip(self), err, ret)]
    pub fn prepare(&mut self, query: QueryString) -> Result<QueryResult, Error> {
        self.prepare_with_id(query, ulid::Ulid::new().0)
    }

    pub fn prepare_with_id(&mut self, query: QueryString, id: u128) -> Result<QueryResult, Error> {
        if !matches!(
            query,
            QueryString::Insert { .. } | QueryString::Delete { .. } | QueryString::Select { .. }
        ) {
            return Err(Error::new(
                DbError::Invalid,
                "Only INSERT, DELETE or SELECT can be prepared",
            ));
        }
        let Some(keyspace) = query.keyspace().or(self.use_keyspace.as_deref()) else {
            Err(Error::new(DbError::Invalid, "keyspace must be specified"))?
        };
        let Some(keyspace) = self.keyspace.get_mut(keyspace) else {
            Err(Error::new(DbError::Invalid, "keyspace does not exist"))?
        };
        let (prepared_metadata, result_metadata) = keyspace.prepare(query.clone())?;

        self.prepared.insert(id, query);

        let prepared = Prepared {
            id,
            prepared_metadata,
            result_metadata,
        };

        Ok(QueryResult::Prepared(prepared))
    }

    pub fn use_keyspace(&mut self, ks: impl Into<String>) {
        self.use_keyspace = Some(ks.into());
    }
}
