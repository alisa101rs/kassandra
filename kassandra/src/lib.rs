pub mod cql;
pub mod error;
pub mod frame;
pub mod session;
pub mod snapshot;
pub mod storage;

pub use session::KassandraSession;
