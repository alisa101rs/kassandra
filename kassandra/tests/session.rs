use insta::assert_debug_snapshot;
use kassandra::{
    frame::{request::query::Query, response::result::QueryResult},
    KassandraSession,
};

macro_rules! exec {
    ($session: tt, $query: tt) => {
        $session.process(Query::simple($query).unwrap()).unwrap()
    };
}

pub fn session() -> KassandraSession {
    let mut session = KassandraSession::new();
    let result = exec!(
        session,
        "CREATE KEYSPACE cycling
          WITH REPLICATION = {
           'class' : 'SimpleStrategy',
           'replication_factor' : 1
          };"
    );

    assert!(matches! {result, QueryResult::SchemaChange(_)});

    let result = exec!(
        session,
        "CREATE TABLE cycling.cyclist_name (
                       id int PRIMARY KEY,
                       lastname text,
                       firstname text,
                       records map<text, text>);"
    );
    assert!(matches! {result, QueryResult::SchemaChange(_)});
    session
}

#[test]
fn create_session_and_table() {
    let _ = session();
}

#[test]
fn scan_simple_data() {
    let mut session = session();
    let result = exec!(
        session,
        "insert into cycling.cyclist_name (id, lastname, firstname, records) values (1, 'john', 'johnson', {'f1': '120', 'f2': '126'});"
    );
    assert!(matches! { result, QueryResult::Void});
    let QueryResult::Rows(rows) = exec!(session, "select * from cycling.cyclist_name;") else {
        panic!("invalid return type");
    };
    assert_debug_snapshot!("select all", rows);

    let QueryResult::Rows(rows) = exec!(
        session,
        "select id, firstname as name, toJson(records) as rec from cycling.cyclist_name;"
    ) else {
        panic!("invalid return type");
    };
    assert_debug_snapshot!("select explicit columns", rows);

    let QueryResult::Rows(rows) = exec!(session, "select json * from cycling.cyclist_name;") else {
        panic!("invalid return type");
    };
    assert_eq!(rows.metadata.col_specs.len(), 1);
    assert_debug_snapshot!("select json", rows);
}

#[test]
fn select_simple_data() {
    let mut session = session();
    let _ = exec!(
        session,
        "insert into cycling.cyclist_name (id, lastname, firstname, records) values (1, 'john', 'johnson', {'f1': '120', 'f2': '126'});"
    );
    let _ = exec!(
        session,
        "insert into cycling.cyclist_name (id, lastname, firstname, records) values (2, 'smith', 'smithson', {'f1': '120', 'f2': '126'});"
    );

    let QueryResult::Rows(rows) = exec!(
        session,
        "select firstname as name, toJson(records) as rec from cycling.cyclist_name where id = 2;"
    ) else {
        panic!("invalid return type");
    };

    assert_debug_snapshot!("select single row", rows);
}
