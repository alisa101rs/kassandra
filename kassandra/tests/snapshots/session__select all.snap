---
source: kassandra/tests/session.rs
expression: rows
---
Rows {
    metadata: ResultMetadata {
        global_spec: Some(
            TableSpec {
                ks_name: "cycling",
                table_name: "cyclist_name",
            },
        ),
        paging_state: None,
        col_specs: [
            ColumnSpec {
                table_spec: None,
                name: "firstname",
                typ: Text,
            },
            ColumnSpec {
                table_spec: None,
                name: "id",
                typ: Int,
            },
            ColumnSpec {
                table_spec: None,
                name: "lastname",
                typ: Text,
            },
            ColumnSpec {
                table_spec: None,
                name: "records",
                typ: Map(
                    Text,
                    Text,
                ),
            },
        ],
    },
    rows: [
        Row {
            columns: [
                Some(
                    Text(
                        "johnson",
                    ),
                ),
                Some(
                    Int(
                        1,
                    ),
                ),
                Some(
                    Text(
                        "john",
                    ),
                ),
                Some(
                    Map(
                        [
                            (
                                Text(
                                    "f1",
                                ),
                                Text(
                                    "120",
                                ),
                            ),
                            (
                                Text(
                                    "f2",
                                ),
                                Text(
                                    "126",
                                ),
                            ),
                        ],
                    ),
                ),
            ],
        },
    ],
}
